R/cnd-message.R

```diff
@@ -1,121 +1,126 @@
-#' Build an error message from a main issue and bullet messages
+#' Build an error message from parts
 #'
 #' @description
 #'
-#' \Sexpr[results=rd, stage=render]{rlang:::lifecycle("experimental")}
-#'
-#' `cnd_message()` assembles an error message from two components:
+#' `cnd_message()` assembles an error message from three generics:
 #'
-#' - The `cnd_issue()` generic. Methods should return a single line.
+#' - `cnd_header()`
+#' - `cnd_body()`
+#' - `cnd_footer()`
 #'
-#' - The `cnd_bullets()` generic. Methods should return a named vector
-#'   of lines. These lines are automatically prefixed with a bullet by
-#'   `cnd_message()` (see the section on error statements).
+#' The default method for the error header returns the `message` field
+#' of the condition object. The default methods for the body and
+#' footer return empty character vectors.
 #'
 #' `cnd_message()` is automatically called by the `conditionMessage()`
-#' for rlang errors so that errors thrown with [abort()] only need to
-#' implement `cnd_issue()` and `cnd_bullets()`. It can also be called
-#' in custom `conditionMessage()` methods.
-#'
-#' Note that if you pass a named character vector to [abort()], you
-#' get the same formatting behaviour as `cnd_message()`.
-#'
-#' @param cnd A condition object.
-#'
-#' @section Error statements:
-#'
-#' This experimental infrastructure is based on the idea that
-#' sentences in error messages are best kept short and simple. From
-#' this point of view, the best way to present the information is as a
-#' bullet list of simple sentences containing a single clause.
-#' `cnd_message()` helps following this structure by building an error
-#' message from two parts: the __issue__ and the __bullets__.
-#'
-#' `cnd_issue()` is the generic for the main error message. It should
-#' be as generic as possible, but since it is a generic it is easy to
-#' override by error subclasses.
+#' for rlang errors. Error classes created with [abort()] only need to
+#' implement header, body or footer methods.
 #'
-#' The `cnd_bullets()` methods should return a character vector of
-#' sentences. These are automatically prefixed with bullets by
-#' `cnd_message()`, according to the following scheme:
 #'
-#' - Elements named `"i"` are prefixed with a blue "info" symbol.
-#' - Elements named `"x"` are prefixed with a red "cross" symbol.
-#' - Unnamed elements are prefixed with a "*" symbol.
+#' @section Overriding `cnd_body()`:
 #'
-#' While you are free to lay out the bullets in the order that you
-#' like, "x" bullets should usually precede "i" bullets.
+#' Sometimes the contents of an error message depends on the state of
```

Comment:
This idea slightly freaks me out, but it seems to makes sense.

Reply:
This overriding was how we could achieve lazy errors + complicated type checking + shallow error hierachy in vctrs.

In general I think the idea of extending S3 to be truly OO (i.e. the object itself decides its implementation, rather than the lexical scoping or the subclassing hierarchy) is worth exploring.

Reply:
This section now starts with an `experimental` badge.